packages/core/src/components/button/buttons.tsx

```diff
@@ -41,41 +29,50 @@ export class Button extends React.Component<React.HTMLProps<HTMLButtonElement> &
     public static displayName = "Blueprint.Button";
 
     public render() {
-        return (
-            <button
-                type="button"
-                {...removeNonHTMLProps(this.props)}
-                className={getButtonClasses(this.props)}
-                ref={this.props.elementRef}
-            >
-                {this.props.text}
-                {this.props.children}
-                {maybeRenderRightIcon(this.props.rightIconName)}
-            </button>
-        );
+        return <button type="button" {...getButtonHTMLProps(this.props)} {...getButtonProps(this.props)} />;
     }
 }
 
 export class AnchorButton extends React.Component<React.HTMLProps<HTMLAnchorElement> & IButtonProps, {}> {
     public static displayName = "Blueprint.AnchorButton";
 
     public render() {
-        const { children, disabled, href, onClick, rightIconName, text } = this.props;
-        return (
-            <a
-                {...removeNonHTMLProps(this.props)}
-                className={getButtonClasses(this.props)}
-                href={disabled ? null : href}
-                onClick={disabled ? null : onClick}
-                ref={this.props.elementRef}
-                role="button"
-                tabIndex={disabled ? null : 0}
-            >
-                {text}
-                {children}
-                {maybeRenderRightIcon(rightIconName)}
-            </a>
-        );
+        return <a role="button" {...getButtonHTMLProps(this.props)} {...getButtonProps(this.props)} />;
+    }
+}
+
+function getButtonHTMLProps(props: IButtonProps) {
+    return removeNonHTMLProps(props, props.disabled ? DISABLED_PROPS : [], true);
+}
+
+// spread this after HTML props to override `className`
+function getButtonProps(props: IButtonProps & { children?: React.ReactNode }) {
+    return {
+        children: [
+            props.text,
+            props.children,
+            maybeRenderRightIcon(props.rightIconName),
+        ],
+        className: getButtonClasses(props),
+        ref: props.elementRef,
+    };
+}
+
+function getButtonClasses(props: IButtonProps) {
+    return classNames(
+        Classes.BUTTON,
+        { [Classes.DISABLED]: props.disabled },
+        Classes.iconClass(props.iconName),
+        Classes.intentClass(props.intent),
+        props.className
+    );
+}
+
+function maybeRenderRightIcon(iconName: string) {
+    if (iconName == null) {
+        return null;
```

Comment:
`maybeRender...` methods shouldn't return `null` to represent the lack of a value; they should use `undefined`. the only time we are forced to use `null` in contexts like this is the top-level `render` method of a component (it can't return `undefined`.